### PR TITLE
fix: [CDS-77251]: support for readonly fields for servicenow standard templates

### PR DESCRIPTION
**Pull request by vivekkarnatiharness**
## Describe your changes
image

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/52419)

### CODE REVIEW

**Comment by vivekkarnatiharness** [comment]
trigger uts

File: 125-cd-nextgen/src/main/java/io/harness/cdng/servicenow/resources/service/ServiceNowResourceServiceImpl.java [hunk 1]
```diff
@@ -235,6 +238,29 @@ public ServiceNowTicketNG getTicketDetails(IdentifierRef connectorRef, String or
...
+          ServiceNowTaskNGParameters.builder().action(ServiceNowActionNG.GET_STANDARD_TEMPLATES_READONLY_FIELDS);
+      String serviceNowStandardTemplateReadOnlyFields =
+          obtainServiceNowTaskNGResponse(connectorRef, orgId, projectId, parametersBuilder)
+              .getServiceNowStandardTemplateReadOnlyFields();
+      if (StringUtils.isBlank(serviceNowStandardTemplateReadOnlyFields)) {
+        return Collections.emptyList();
+      }
+      return List.of(serviceNowStandardTemplateReadOnlyFields.split(","));
+    } catch (ServiceNowException e) {
+      if (e.getMessage().contains("User is unauthorized to access table: std_change_properties")) {
```
Comment (naman-goenka) [inline 125-cd-nextgen/src/main/java/io/harness/cdng/servicenow/resources/service/ServiceNowResourceServiceImpl.java:254]: npe possible here

File: 125-cd-nextgen/src/main/java/io/harness/cdng/servicenow/resources/service/ServiceNowResourceServiceImpl.java [hunk 2]
```diff
@@ -235,6 +238,29 @@ public ServiceNowTicketNG getTicketDetails(IdentifierRef connectorRef, String or
...
+        return Lists.newArrayList("description", "backout_plan", "test_plan", "implementation_plan");
```
Comment (naman-goenka) [inline 125-cd-nextgen/src/main/java/io/harness/cdng/servicenow/resources/service/ServiceNowResourceServiceImpl.java:255]: We can covert this to a static variable for readability

File: 125-cd-nextgen/src/main/java/io/harness/cdng/servicenow/resources/service/ServiceNowResourceServiceImpl.java (hunk 1 above)
Comment (naman-goenka) [reply 125-cd-nextgen/src/main/java/io/harness/cdng/servicenow/resources/service/ServiceNowResourceServiceImpl.java:254]: Please log the exception as well ex: "Acl related error occured while fetching readonly fields: .."

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/servicenow/ServiceNowTaskNgHelper.java [hunk 3]
```diff
@@ -667,7 +669,44 @@ private ServiceNowTaskNGResponse getMetadataV2(ServiceNowTaskNGParameters servic
...
+    ServiceNowConnectorDTO serviceNowConnectorDTO = serviceNowTaskNGParameters.getServiceNowConnectorDTO();
+    ServiceNowRestClient serviceNowRestClient = getServiceNowRestClient(serviceNowConnectorDTO.getServiceNowUrl());
 
+    final Call<JsonNode> request = serviceNowRestClient.getReadOnlyFieldsForStandardTemplate(
+        ServiceNowAuthNgHelper.getAuthToken(serviceNowConnectorDTO));
+    Response<JsonNode> response = null;
+    try {
+      response = Retry.decorateCallable(retry, () -> request.clone().execute()).call();
+      log.info("Response received from serviceNow for GET_STANDARD_TEMPLATES_READONLY_FIELDS: {}", response);
+      handleResponse(response, "Failed to get serviceNow fields");
```
Comment (naman-goenka) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/task/servicenow/ServiceNowTaskNgHelper.java:683]: Please update this "Failed to get serviceNow fields"

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/servicenow/ServiceNowTaskNgHelper.java [hunk 4]
```diff
@@ -667,7 +669,44 @@ private ServiceNowTaskNGResponse getMetadataV2(ServiceNowTaskNGParameters servic
...
+      JsonNode responseObj = response.body().get("result");
+      if (responseObj != null) {
+        if (responseObj.get(0) == null || responseObj.get(0).get("readonly_fields") == null) {
```
Comment (naman-goenka) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/task/servicenow/ServiceNowTaskNgHelper.java:686]: nit: we can do this parsing as well or manager side

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/servicenow/ServiceNowTaskNgHelper.java (hunk 3 above)
Comment (vivekkarnatiharness) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/task/servicenow/ServiceNowTaskNgHelper.java:683]: I think i used the same string.Is this by mistake??

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/servicenow/ServiceNowTaskNgHelper.java (hunk 4 above)
Comment (vivekkarnatiharness) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/task/servicenow/ServiceNowTaskNgHelper.java:686]: I think passing json object as a string to manager, will make manager to again convert json to string.I think its better to parse this small logic on delegate itself

**Comment by vivekkarnatiharness** [comment]
trigger uts

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/servicenow/ServiceNowTaskNgHelper.java (hunk 3 above)
Comment (naman-goenka) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/task/servicenow/ServiceNowTaskNgHelper.java:683]: This is used to construct expression please replace with Failed to get readonly metadata ...

**Comment by vivekkarnatiharness** [comment]
trigger uts

**Review by abhinavmittal-harness — CHANGES_REQUESTED** [review]
> Reviewed 2 of 4 files at r2, 1 of 2 files at r3, 4 of 4 files at r5, 2 of 2 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/52419)* status: 9 file(s) to review. 1 change requests. 1. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/servicenow/resources/service/ServiceNowResourceServiceImpl.java` line 259 at r6](https://reviewable.io/reviews/harness/harness-core/52419#-NeDNStzCRcRpxFrzvrF:-NeDNStzCRcRpxFrzvrG:b-xjo2ls) ([raw file](https://github.com/harness/harness-core/blob/0f90f6c84f83b76e789a228f82ed1b827d6c2a09/125-cd-nextgen/src/main/java/io/harness/cdng/servicenow/resources/service/ServiceNowResourceServiceImpl.java#L259)):*
> ```Java
>       if (e.getMessage() != null
>           && e.getMessage().contains("User is unauthorized to access table: std_change_properties")) {
>         log.error("Acl related error occured while fetching readonly fields:", e.getMessage());
> ```

It is almost always better to send the complete exception, so that stack trace can be logged.

**Review by vivekkarnatiharness — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/52419)* status: 9 file(s) to review. 1 change requests. 4. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/servicenow/resources/service/ServiceNowResourceServiceImpl.java` line 254 at r3](https://reviewable.io/reviews/harness/harness-core/52419#gh-1324274714:-NeD8Igj2miLvYCqJKWd:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/17c2c6cf03c74420c7cd2759f5250f68790c7a06/125-cd-nextgen/src/main/java/io/harness/cdng/servicenow/resources/service/ServiceNowResourceServiceImpl.java#L254)):*
<details><summary><i>Previously, naman-goenka (Naman Goenka) wrote…</i></summary><blockquote>

Please log the exception as well ex: "Acl related error occured while fetching readonly fields: .."
</blockquote></details>

Done.
___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/servicenow/resources/service/ServiceNowResourceServiceImpl.java` line 255 at r3](https://reviewable.io/reviews/harness/harness-core/52419#gh-1324275630:-NeD8IL71GKk1NmYgRI4:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/17c2c6cf03c74420c7cd2759f5250f68790c7a06/125-cd-nextgen/src/main/java/io/harness/cdng/servicenow/resources/service/ServiceNowResourceServiceImpl.java#L255)):*
<details><summary><i>Previously, naman-goenka (Naman Goenka) wrote…</i></summary><blockquote>

We can convert this to a static variable for readability
</blockquote></details>

Done.
___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/servicenow/resources/service/ServiceNowResourceServiceImpl.java` line 259 at r6](https://reviewable.io/reviews/harness/harness-core/52419#-NeDNStzCRcRpxFrzvrF:-NeDScD5EK1LBIBq5Zaj:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/0f90f6c84f83b76e789a228f82ed1b827d6c2a09/125-cd-nextgen/src/main/java/io/harness/cdng/servicenow/resources/service/ServiceNowResourceServiceImpl.java#L259)):*
<details><summary><i>Previously, abhinavmittal-harness (Abhinav Mittal) wrote…</i></summary><blockquote>

It is almost always better to send the complete exception, so that stack trace can be logged.
</blockquote></details>

Done.

**Review by abhinavmittal-harness — COMMENTED** [review]
> Reviewed 1 of 1 files at r8, 1 of 1 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/52419)* status: 9 file(s) to review. 1 change requests. 4. unresolved discussions. 0 of 2 approvals obtained.

**Review by vivekkarnatiharness — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/52419)* status: 9 file(s) to review. 1 change requests. 5. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`930-delegate-tasks/src/main/java/io/harness/delegate/task/servicenow/ServiceNowTaskNgHelper.java` line 683 at r3](https://reviewable.io/reviews/harness/harness-core/52419#gh-1324284207:-NeDlObzCfsMHQsywcAe:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/17c2c6cf03c74420c7cd2759f5250f68790c7a06/930-delegate-tasks/src/main/java/io/harness/delegate/task/servicenow/ServiceNowTaskNgHelper.java#L683)):*
<details><summary><i>Previously, naman-goenka (Naman Goenka) wrote…</i></summary><blockquote>

This is used to construct expression please replace with Failed to get readonly metadata ...
</blockquote></details>

Done.

**Review by naman-goenka — APPROVED** [review]

**Comment by vivekkarnatiharness** [comment]
trigger uts

**Comment by vivekkarnatiharness** [comment]
trigger review

**Pull request merged**